File: food-ordering-system/restaurant-service/src/main/java/mk/finki/restaurantservice/mcp/RestaurantMcpServer.java
package mk.finki.restaurantservice.mcp;

import lombok.RequiredArgsConstructor;
import mk.finki.restaurantservice.application.RestaurantApplicationService;
import mk.finki.restaurantservice.domain.Restaurant;
import org.springframework.stereotype.Component;
import java.util.List;
import java.util.UUID;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * A simplified MCP Server implementation for demonstration.
 * In a real scenario, this would use an MCP SDK or expose a JSON-RPC over stdio/HTTP.
 */
@Component
@RequiredArgsConstructor
public class RestaurantMcpServer {
    private final RestaurantApplicationService restaurantService;

    // MCP Tools & Resources
    
    public List<Map<String, Object>> listRestaurants() {
        return restaurantService.listRestaurants().stream()
                .map(r -> Map.of("id", r.getId(), "name", r.getName()))
                .collect(Collectors.toList());
    }

    public Map<String, Object> getRestaurantMenu(UUID restaurantId) {
        Restaurant r = restaurantService.getRestaurant(restaurantId);
        return Map.of(
            "restaurantName", r.getName(),
            "menu", r.getMenu().stream()
                .map(item -> Map.of(
                    "id", item.getId(),
                    "name", item.getName(),
                    "price", item.getPrice(),
                    "status", item.getStatus()
                )).collect(Collectors.toList())
        );
    }
}

File: food-ordering-system/restaurant-service/src/main/kotlin/mk/finki/restaurantservice/mcp/RestaurantMcpServer.kt
package mk.finki.restaurantservice.mcp

import mk.finki.restaurantservice.application.MenuItemApplicationService
import mk.finki.restaurantservice.application.RestaurantApplicationService
import org.springframework.stereotype.Component
import java.util.*

/**
 * MCP Server implementation for Restaurant Service.
 * Exposes read-only tools to AI clients via the Model Context Protocol.
 * Justification for embedding: Avoids duplication of DTOs and provides direct access to Application Services.
 */
@Component
class RestaurantMcpServer(
    private val restaurantService: RestaurantApplicationService,
    private val menuItemService: MenuItemApplicationService
) {
    // Mocking the MCP standard registration/loop for demonstration in docs
    
    fun listRestaurants(): String {
        val restaurants = restaurantService.listRestaurants()
        return restaurants.joinToString("\n") { "${it.id.value}: ${it.name} at ${it.address}" }
    }

    fun getRestaurantMenu(restaurantId: UUID): String {
        val restaurant = restaurantService.getRestaurant(restaurantId)
            .orElseThrow { RuntimeException("Restaurant not found") }
        return restaurant.menu.joinToString("\n") { "- ${it.name}: ${it.price.amount} ${it.price.currency} (${it.status})" }
    }

    fun checkItemAvailability(ids: List<UUID>): String {
        val availability = menuItemService.checkAvailability(ids)
        return availability.joinToString("\n") { "${it.menuItemId}: ${if (it.available) "Available" else "Unavailable"}" }
    }
}
